lib/pbench/server/api/resources/__init__.py

```diff
@@ -696,6 +697,203 @@ def convert_boolean(value: Any, parameter: "Parameter") -> bool:
+
+    def parse_expression(self) -> "Term":
+        """Parse a filter expression like "<key>:[<operator>]<value>[:<type>]"
+
+        Separates the "key", "operator", "value", and "type" fields of a filter
+        expression.
+
+        The key and value can be quoted to include mixed case and symbols,
+        including the ":" character, by starting and starting and ending the
+        key or value with the "'" or '"' character, e.g.,
+        "'dataset.metalog.tool/iteration:1':'foo:1#bar':str".
+
+        The "operator" and "type", if omitted, can be defaulted from the object
+        configuration, so that `dataset.name:foo` can have the same effect as
+        `dataset.name:=foo:str`.
+
+        Returns:
+            self
+        """
+
+        # The "chain" allows chaining multiple expressions as an OR
+        self.chain = self._remove_prefix(("^",))
+
+        # The metadata key token
+        self.key = self._next_token()
+        if not Metadata.is_key_path(
+            self.key, self.keys, metalog_key_ok=bool(self.operators)
+        ):
+            raise APIAbort(HTTPStatus.BAD_REQUEST, str(MetadataBadKey(self.key)))
+
+        # The comparison operator (defaults to "=")
+        if self.operators:
+            self.operator = self._remove_prefix(
+                self.operators.keys(), default=self.default_operator
+            )
+
+        # The filter value
+        self.value = self._next_token(optional=True)
+
+        # The comparison type, defaults to "str"
+        if self.types:
+            self.type = self.buffer.lower() if self.buffer else self.default_type
+            if self.type and self.type not in self.types:
+                raise APIAbort(
+                    HTTPStatus.BAD_REQUEST,
+                    f"The metadata type {self.type!r} must be one of {','.join(t for t in self.types.keys())}",
+                )
+        if not self.key or not self.value:
+            raise APIAbort(
+                HTTPStatus.BAD_REQUEST,
+                f"Metadata {self.term!r} must have the form 'k:v[:type]'",
+            )
+        return self
+
+    def _remove_prefix(
+        self, prefixes: tuple[str], default: Optional[str] = None
+    ) -> Optional[str]:
+        """Remove a string prefix.
+
+        Looking at the current buffer, remove a known prefix before processing
+        the next character. A list of prefixes may be specified: they're sorted
+        longest first so that we don't accidentally match a substring.
+
+        Args:
+            prefixes: a collection of prefix strings
+            default: the default prefix if none appears in the string.
+
+        Returns:
+            The prefix, or the default.
+        """
+        prefix = default
+        for p in sorted(prefixes, key=lambda k: len(k), reverse=True):
+            if self.buffer.startswith(p):
+                prefix = self.buffer[: len(p)]
+                self.buffer = self.buffer[len(p) :]
+                self.offset += len(p)
```

Comment:
The tokenizer works in a moving `buffer`, but I decided I wanted to generate messages using the full "context" rather than just the span of the current token. Therefore I need to figure out where the current moving buffer is relative to the original: keeping the position was the easiest way. Another would be to work solely from `self.buffer` and `self.offset` *without* moving `self.buffer` ... I didn't feel like doing that.

Reply:
I think that moving `self.buffer` confers various simplifications, so I would keep that.

More context is probably better, and there aren't _that_ many places which maintain `self.offset`, so...nevermind.